File: backend/app/Models/Ticket.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Str;

class Ticket extends Model
{
    protected $fillable = [
        'violator_id',
        'enforcer_id',
        'violation_id',
        'fine_amount',
        'additional_fees',
        'total_amount',
        'place_of_apprehension',
        'apprehended_at',
        'control_no',
        'status',
    ];

    protected $casts = [
        'apprehended_at'   => 'datetime',
        'fine_amount'      => 'decimal:2',
        'additional_fees'  => 'decimal:2',
        'total_amount'     => 'decimal:2',
    ];

    // 👇 this tells Laravel to include `enforcer_name` in JSON
    protected $appends = ['enforcer_name'];

    /**
     * Auto-generate control_no if not provided.
     */
    protected static function booted()
    {
        static::creating(function (Ticket $ticket) {
            if (empty($ticket->control_no)) {
                $ticket->control_no = static::nextControlNo();
            }
        });
    }

    /**
     * Pattern: YYYYMMDD-#### e.g. 20251130-0001
     */
    public static function nextControlNo(): string
    {
        $today  = now()->format('Ymd');
        $prefix = $today . '-';

        $last = static::where('control_no', 'like', $prefix . '%')
            ->orderByDesc('control_no')
            ->first();

        $nextSeq = 1;

        if ($last) {
            $lastSeq = (int) Str::afterLast($last->control_no, '-');
            $nextSeq = $lastSeq + 1;
        }

        return sprintf('%s%04d', $prefix, $nextSeq);
    }

    public function violator(): BelongsTo
    {
        return $this->belongsTo(Violator::class);
    }

    public function enforcer(): BelongsTo
    {
        return $this->belongsTo(User::class, 'enforcer_id');
    }

    public function violation(): BelongsTo
    {
        return $this->belongsTo(Violation::class);
    }

    public function violations()
    {
        // in case you later use multiple violations
        return $this->belongsToMany(Violation::class, 'ticket_violation')
            ->withPivot('fine_amount', 'remarks')
            ->withTimestamps();
    }

    public function payments(): HasMany
    {
        return $this->hasMany(Payment::class);
    }

    // 👇 THIS is what Flutter will read as `enforcer_name`
    public function getEnforcerNameAttribute()
    {
        // users table uses `full_name`, not `name`
        return $this->enforcer ? $this->enforcer->full_name : null;
    }
}

File: lib/models/ticket.dart
class Ticket {
  final int id;
  final int violatorId;
  final int enforcerId;
  final int? violationId; // if single violation per ticket
  final double fineAmount;
  final double additionalFees;
  final double totalAmount;
  final String placeOfApprehension;
  final DateTime apprehendedAt;
  final String controlNo;
  final String status;

  Ticket({
    required this.id,
    required this.violatorId,
    required this.enforcerId,
    this.violationId,
    required this.fineAmount,
    required this.additionalFees,
    required this.totalAmount,
    required this.placeOfApprehension,
    required this.apprehendedAt,
    required this.controlNo,
    required this.status,
  });

  factory Ticket.fromJson(Map<String, dynamic> json) {
    return Ticket(
      id: json['id'],
      violatorId: json['violator_id'],
      enforcerId: json['enforcer_id'],
      violationId: json['violation_id'],
      fineAmount: double.tryParse(json['fine_amount'].toString()) ?? 0.0,
      additionalFees:
          double.tryParse(json['additional_fees'].toString()) ?? 0.0,
      totalAmount: double.tryParse(json['total_amount'].toString()) ?? 0.0,
      placeOfApprehension: json['place_of_apprehension'] ?? '',
      apprehendedAt: DateTime.parse(json['apprehended_at']),
      controlNo: json['control_no'],
      status: json['status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'violator_id': violatorId,
      'enforcer_id': enforcerId,
      'violation_id': violationId,
      'fine_amount': fineAmount,
      'additional_fees': additionalFees,
      'total_amount': totalAmount,
      'place_of_apprehension': placeOfApprehension,
      'apprehended_at': apprehendedAt.toIso8601String(),
      'control_no': controlNo,
      'status': status,
    };
  }
}
